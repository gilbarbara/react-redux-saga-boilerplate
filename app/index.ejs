<!DOCTYPE html>
<html lang="en" class=""
  <% if(htmlWebpackPlugin.files.manifest) { %>
      manifest="<%= htmlWebpackPlugin.files.manifest %>"
  <% } %>
>
<head>
  <meta charset="utf-8">
  <title><%= htmlWebpackPlugin.options.title || 'Webpack App' %></title>
  <% if (htmlWebpackPlugin.options.mobile) { %>
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1.0, user-scalable=0">
  <% } %>
  <meta property="og:title" content="React-Redux-Saga Boilerplate" />
  <meta property="og:url" content="https://gilbarbara.github.io/react-redux-saga-boilerplate/" />
  <meta property="og:description" content="Provides: react ^15.x, react-router ^2.x, redux ^3.x, redux-saga ^0.11" />
  <meta property="og:image" content="https://gilbarbara.github.io/react-redux-saga-boilerplate/media/og-image-v1.png" />
  <meta property="og:type" content="website" />
  <% require('assets/media/images/og-image-v1.png') %>
  <link rel="shortcut icon" href="<%- require('assets/media/icons/favicon.ico') %>">
  <% for (var css in htmlWebpackPlugin.files.css) { %>
  <link href="<%= htmlWebpackPlugin.files.css[css] %>" rel="stylesheet">
  <% } %>
  <meta name="apple-mobile-web-app-title" content="React-Redux-Saga">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="#333333">
  <link rel="apple-touch-icon" sizes="180x180" href="<%- require('assets/media/icons/apple-touch-icon.png') %>">
  <link rel="icon" type="image/png" href="<%- require('assets/media/icons/favicon-32x32.png') %>" sizes="32x32">
  <link rel="icon" type="image/png" href="<%- require('assets/media/icons/favicon-16x16.png') %>" sizes="16x16">
  <link rel="manifest" href="<%- require('assets/manifest.json') %>">
  <% require('assets/media/icons/icon-96x96.png') %>
  <% require('assets/media/icons/icon-144x144.png') %>
  <% require('assets/media/icons/icon-192x192.png') %>
  <% require('assets/media/icons/icon-512x512.png') %>
  <link rel="mask-icon" href="<%- require('assets/media/icons/safari-pinned-tab.svg') %>" color="#5bbad5">
  <meta name="theme-color" content="#333333">
  <% if (htmlWebpackPlugin.options.baseHref) { %>
  <base href="<%= htmlWebpackPlugin.options.baseHref %>" />
  <% } %>
  <style type="text/css">.app { opacity: 0 }</style>
</head>
<body>

<div id="react" style="min-height: 100vh;"></div>

<% if (htmlWebpackPlugin.options.window) { %>
<script>
  <% for (var varName in htmlWebpackPlugin.options.window) { %>
  window['<%= varName %>'] = <%= JSON.stringify(htmlWebpackPlugin.options.window[varName]) %>;
  <% } %>
</script>
<% } %>

<% for (var chunk in htmlWebpackPlugin.files.chunks) { %>
<script src="<%= htmlWebpackPlugin.files.chunks[chunk].entry %>"></script>
<% } %>

<% if (htmlWebpackPlugin.options.googleAnalytics) { %>
<script>
  (function(i, s, o, g, r, a, m) {
    i['GoogleAnalyticsObject'] = r;
    i[r] = i[r] || function() {
        (i[r].q = i[r].q || []).push(arguments)
      }, i[r].l = 1 * new Date();
    a = s.createElement(o),
      m = s.getElementsByTagName(o)[0];
    a.async = 1;
    a.src = g;
    m.parentNode.insertBefore(a, m)
  })(window, document, 'script', '//www.google-analytics.com/analytics.js', 'ga');


  <% if (htmlWebpackPlugin.options.googleAnalytics.trackingId) { %>
  ga('create', '<%= htmlWebpackPlugin.options.googleAnalytics.trackingId %>', 'auto');
  <% } else {
    throw new Error("html-webpack-template requires googleAnalytics.trackingId config");
  } %>

  <% if (htmlWebpackPlugin.options.googleAnalytics.pageViewOnLoad) { %>
  ga('send', 'pageview');
  <% } %>
</script>
<% } %>
</body>
</html>
