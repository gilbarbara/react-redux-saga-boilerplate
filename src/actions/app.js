// @flow
/**
 * @module Actions/App
 * @desc App Actions
 */

import uuid from 'uuid/v4';
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export { goBack, go, push, replace } from 'modules/history';

export const { hideAlert, showAlert, switchMenu } = createActions({
  [ActionTypes.SWITCH_MENU]: (query: string) => ({ query }),
  [ActionTypes.HIDE_ALERT]: (id: string) => ({ id }),
  [ActionTypes.SHOW_ALERT]: (message: string, options: Object) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || uuid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
});
