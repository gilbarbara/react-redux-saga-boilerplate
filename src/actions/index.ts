export * from './app';
export * from './github';
export * from './user';
