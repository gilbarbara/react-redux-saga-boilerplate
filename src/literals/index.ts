import { keyMirror } from 'modules/helpers';

import { Status } from 'types';

export const ActionTypes = keyMirror({
  GITHUB_GET_REPOS_REQUEST: undefined,
  GITHUB_GET_REPOS_SUCCESS: undefined,
  GITHUB_GET_REPOS_FAILURE: undefined,
  HIDE_ALERT: undefined,
  SHOW_ALERT: undefined,
  USER_LOGIN_REQUEST: undefined,
  USER_LOGIN_SUCCESS: undefined,
  USER_LOGIN_FAILURE: undefined,
  USER_LOGOUT_REQUEST: undefined,
  USER_LOGOUT_SUCCESS: undefined,
  USER_LOGOUT_FAILURE: undefined,
});

export const STATUS: Status = {
  IDLE: 'idle',
  RUNNING: 'running',
  READY: 'ready',
  SUCCESS: 'success',
  ERROR: 'error',
};
