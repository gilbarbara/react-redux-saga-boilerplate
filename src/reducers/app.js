import { handleActions } from 'redux-actions';
import immutable from 'immutability-helper';
import { REHYDRATE } from 'redux-persist/lib/constants';

import { ActionTypes } from 'constants/index';

export const appState = {
  alerts: [],
};

export default {
  app: handleActions(
    {
      [REHYDRATE]: state =>
        immutable(state, {
          alerts: { $set: [] },
        }),
      [ActionTypes.HIDE_ALERT]: (state, { payload: { id } }) => {
        const alerts = state.alerts.filter(d => d.id !== id);

        return immutable(state, {
          alerts: { $set: alerts },
        });
      },
      [ActionTypes.SHOW_ALERT]: (state, { payload }) =>
        immutable(state, {
          alerts: { $push: [payload] },
        }),
    },
    appState,
  ),
};
