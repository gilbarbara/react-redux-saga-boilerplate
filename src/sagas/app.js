/**
 * @module Sagas/App
 * @desc App
 */
import { all, put, select, takeLatest } from 'redux-saga/effects';

import { ActionTypes } from 'constants/index';

/**
 * Switch Menu
 *
 * @param {Object} action
 *
 */
export function* switchMenu({ payload }) {
  try {
    const repos = yield select(state => state.github.repos);

    /* istanbul ignore else */
    if (!repos.data[payload.query] || !repos.data[payload.query].length) {
      yield put({
        type: ActionTypes.GITHUB_GET_REPOS,
        payload,
      });
    }
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.EXCEPTION,
      payload: err,
    });
  }
}

/**
 * App Sagas
 */
export default function* root() {
  yield all([takeLatest(ActionTypes.SWITCH_MENU, switchMenu)]);
}
