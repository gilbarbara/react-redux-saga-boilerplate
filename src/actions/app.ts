/**
 * @module Actions/App
 * @desc App Actions
 */
import React from 'react';
import { nanoid } from 'nanoid';

import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { ShowAlertOptions } from 'types';

export { go, goBack, push, replace } from 'modules/history';

export const hideAlert = createAction(ActionTypes.HIDE_ALERT, (id: string) => ({ id }));

export const showAlert = createAction(
  ActionTypes.SHOW_ALERT,
  (message: React.ReactNode, options: ShowAlertOptions) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || nanoid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
);
