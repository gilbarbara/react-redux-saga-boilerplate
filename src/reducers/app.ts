import { REHYDRATE } from 'redux-persist';

import { createReducer } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { AppState } from 'types';

export const appState: AppState = {
  alerts: [],
};

export default {
  app: createReducer<AppState>(
    {
      [REHYDRATE]: draft => {
        draft.alerts = [];
      },
      [ActionTypes.HIDE_ALERT]: (draft, { payload: { id } }) => {
        draft.alerts = draft.alerts.filter(d => d.id !== id);
      },
      [ActionTypes.SHOW_ALERT]: (draft, { payload }) => {
        draft.alerts.push(payload);
      },
    },
    appState,
  ),
};
