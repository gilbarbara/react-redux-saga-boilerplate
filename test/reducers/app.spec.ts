import { hideAlert, showAlert } from 'actions/app';
import { ActionTypes } from 'literals';
import reducer from 'reducers/app';

import { emptyAction } from 'test-utils';

describe('reducers/app', () => {
  let app = reducer.app(undefined, emptyAction);

  it('should return the initial state', () => {
    expect(reducer.app(app, emptyAction)).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.SHOW_ALERT}`, () => {
    app = reducer.app(app, showAlert('HELLO', { id: 'test', variant: 'success' }));
    expect(app).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.HIDE_ALERT}`, () => {
    app = reducer.app(app, hideAlert('test'));
    expect(app).toMatchSnapshot();
  });
});
