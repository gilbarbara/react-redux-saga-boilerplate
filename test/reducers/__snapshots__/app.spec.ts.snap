// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/app should handle HIDE_ALERT 1`] = `
Object {
  "alerts": Array [],
}
`;

exports[`reducers/app should handle SHOW_ALERT 1`] = `
Object {
  "alerts": Array [
    Object {
      "icon": undefined,
      "id": "test",
      "message": "HELLO",
      "position": "bottom-right",
      "timeout": 5,
      "variant": "success",
    },
  ],
}
`;

exports[`reducers/app should return the initial state 1`] = `
Object {
  "alerts": Array [],
}
`;
